fix(legacy-table): avoid React DOM

ktmud opened the pull request:
`jquery.datatables` will manipulate DOMs, sometimes remove them. In case of component being reloaded with updated props, React will not be able to find those removed nodes, causing a `cannot removeChild` error.

Because of the the way to assign row keys, if table shape changes (add or remove columns), React may also have difficulty match the cached nodes via keys.

In general it's a bad idea to directly manipulate React rendered DOM nodes, so we better just let `jquery.datatables` handle everything.

In the future, once we removed `jquery.datatables`, a pure React component will not have such issues.

This PR also contains a fix to null `percent_metrics`.


cc @kristw @etr2460 

🐛 Bug Fix

ktmud committed:
fix(legacy-table): avoid React DOM

`jquery.datatables` will manipulate DOMs, sometimes remove them. In case
of component being reloaded with updated props, React will not be able
to find those removed nodes, causing a `cannot removeChild` error.

Because of the the way to assign row keys, if table shape changes (add
or remove columns), React may also have difficulty match the cached
nodes via keys.

In general it's a bad idea to directly manipulate React rendered DOM
nodes, so we better just let `jquery.datatables` handle everything.

In the future, once we removed `jquery.datatables`, a pure React
component will not have such issues.

ktmud committed:
fix(legacy-table): handle the case when percentMetrics is null

kristw commented:
Can't wait to get rid of `jquery.datatable`. sigh 
Will this change remove the performance gain we had earlier?

ktmud commented:
> Will this change remove the performance gain we had earlier?

I don't think so. `jquery.datatables` will still read an already constructed DOM tree. It's just an extra set `innerHTML` before the React component is mounted.

ktmud committed:
fix(legacy-table): linting errors

kristw commented on scripts/build.js:
```diff
@@ -7,14 +7,18 @@ const glob = process.argv[2];
 };
 
 if (glob) {
+  run(`eslint --color --quiet packages/${glob}/{src,test}/**/*.{js,jsx,ts,tsx}`);
```
`nimbus eslint --color ...`

`nimbus` will create eslint config before executing. Otherwise there may not be eslint config as we don't check them in and let nimbus handle that.

ktmud replied:
I wanted to limit to just the specified packages, but `nimbus` doesn't seem to support that for `eslint` command.

ktmud replied:
My bad. It turns out I can just pass the file directory. Previous tried the same `--workspaces=` approach as other `nimbus` command.

ktmud committed:
refactor: use nimbus build

ktmud committed:
test(legacy-table): add React component tests

ktmud committed:
test(legacy-table): more sophisticated cases

kristw commented on packages/superset-ui-legacy-plugin-chart-table/test/test_data.ts:
```diff
@@ -0,0 +1,103 @@
+/**
```
We adopt camelCase file naming.

ktmud replied:
Thought it was useful to differentiate mock data with actual JS modules. But I guess it doesn't add much value anyway.

kristw commented on packages/superset-ui-legacy-plugin-chart-table/test/ReactDataTable.test.tsx:
```diff
@@ -0,0 +1,65 @@
+import { mount } from 'enzyme';
+import ReactDataTable from '../src/ReactDataTable';
+import transformProps from '../src/transformProps';
+import * as testData from './test_data';
```
Perhaps use specific import to enable warning if we later do not use some of these.
`import { basic, advance } from ...`

ktmud replied:
Wanted to have a namespace so it's more explicit in later code, but also wanted to keep variable names simple.

ktmud replied:
Added a default export for `testData`. I think it looks better than `import *`.

In the future, maybe all plugins can reuse the same test data---similar to switching visualization type in Superset.

kristw commented on packages/superset-ui-legacy-plugin-chart-table/test/ReactDataTable.test.tsx:
```diff
@@ -0,0 +1,65 @@
+  describe('transformProps', () => {});
+
+  describe('ReactDataTable', () => {
+    // Jest throw an error at `console.warn`, this disables that behavior
```
(optional) `nimbus` set this as default, which I also have mixed feelings. I use `jest-mock-console` in another repo, in case the error message if predictable and can be captured.

ktmud replied:
Removed it for now since it's not needed for this plugin anymore.

ktmud committed:
fix: address PR #392 comments

ktmud committed:
chore(legacy-table): clean up tests case setups

Not seeing the console.warn errors anymore. So cleaning it up.

Previously it was from `<SuperChart />` component, but since we have
updated the test case to not use <SuperChart>, we are good now.

ktmud committed:
fix(legacy-table): misleading comment
